src/Cms/Pages.php

```diff
@@ -230,6 +230,15 @@ public function findById(string $id = null)
             return $page;
         }
 
+        // for secondary languages, try the full translated URI
+        if (
+            App::instance()->multilang() === true &&
+            App::instance()->language()->isDefault() === false &&
+            $page = $this->findBy('uri', $id)
+        ) {
+            return $page;
+        }
+
         $start = is_a($this->parent, 'Kirby\Cms\Page') === true ? $this->parent->id() : '';
         $page  = $this->findByIdRecursive($id, $start, App::instance()->multilang());
```

Comment:
```suggestion
        $start = is_a($this->parent, 'Kirby\Cms\Page') === true ? $this->parent->id() : '';
        $page  = $this->findByIdRecursive($id, $start, App::instance()->multilang());

        // for secondary languages, try the full translated URI
        if (
            App::instance()->multilang() === true &&
            App::instance()->language()->isDefault() === false &&
            $page = $this->findBy('uri', $id)
        ) {
            return $page;
        }
```

I would strongly suggest doing the lookup by `uri` _after_ letting the normal lookup occur by `id`. There is an existing oddity that `findById()` actually _prefers_ untranslated versions. Swapping the order should do a better job at keeping this behavior consistent.

See the end of the `testFindByIdWithSwappedSlugsTranslated()` test case lookups. https://github.com/getkirby/kirby/blob/27ae49e07a91323a05c8b87f950fa6ed27bb64ed/tests/Cms/Pages/PagesTest.php#L551-L653

Reply:
Is there a case where the order of the checks matters, i.e. leads to a different result? The existing unit tests all pass, but maybe there's a case that's currently not tested.

When implementing this, I was thinking that all ways of trying to find the page are equal and that they all contribute equally to a non-`null` return value. I.e. only if all ways fail, the result will be `null`.

So with that in mind, I opted for this order because finding by `uri` is faster than the recursive lookup. So for this simple case of "full localized URI is given", the performance will be higher when the code is in this order.

Reply:
Here's an additional test case that could go at the end of `testFindByIdWithSwappedSlugsTranslated()`. It's very similar to the existing last set of tests in that function.

```php
        $pages = new Pages($site->children()->find('aaa', 'aaa/bbb', 'zzz'));
        $this->assertIsPage($pages->findById('aaa'), 'aaa');
        $this->assertIsPage($pages->findById('aaa/bbb'), 'aaa/bbb');
        $this->assertIsPage($pages->findById('zzz'), 'zzz');
        $this->assertIsPage($pages->findById('zzz/yyy'), 'zzz/yyy');
```

In your current proposed logic, under the German test case, the final line will actually return the page that has the `id` of `aaa/bbb`, but if you run the exact same tests under the default language test, it will return `zzz/yyy`.

Reply:
You are absolutely right, thanks for spotting! 👍